flexdll.c

```diff
@@ -356,6 +358,25 @@ static void relocate(resolver f, void *data, reloctbl *tbl, err_t *err) {
       }
       *((UINT32*) ptr->addr) = (INT32) s;
       break;
+    case RELOC_REL32_5:
+      s -= (INT_PTR)(ptr -> addr) + 9;
+      s += *((INT32*) ptr -> addr);
+      if (s != (INT32) s) {
+        sprintf(err->message, "flexdll error: cannot relocate RELOC_REL32_5, target is too far: %p  %p",(void *)((UINT_PTR) s), (void *) ((UINT_PTR)(INT32) s));
+        err->code = 3;
+        goto restore;
+      }
+      *((UINT32*) ptr->addr) = s;
+      break;
+    case RELOC_32NB:
+      s += *((INT32*) ptr -> addr);
+      if (s != (INT32) s) {
+        sprintf(err->message, "flexdll error: cannot relocate %s RELOC_32NB, target is too far: %p  %p", ptr->name, (void *)((UINT_PTR) s), (void *) ((UINT_PTR)(INT32) s));
+        err->code = 3;
+        goto restore;
+      }
+      *((UINT32*) ptr->addr) = s;
```

Comment:
Just for my own understanding:

- In LLVM, this relocation is implemented by `add32(off, s)`, so that `off` in LLVM corresponds to `ptr + addr` in Flexlink and `s` in LLVM corresponds to `s` in Flexlink.
- The relocation `IMAGE_REL_AMD64_ADDR64` in LLVM (`RELOC_ABS` in Flexlink) is implemented by `add64(off, s + imageBase)`, so that `s + imageBase` in LLVM corresponds to `s` in Flexlink.

Do you understand why the `s` arguments do not seem to match between the two cases?

Reply:
Going one step further, those equations imply that `imageBase = 0`.

I spent a lot of time looking and I couldn't find out why IMAGE_REL_AMD64_ADDR64 seems to work. The math for all the other relocation kinds make sense.

However, two things ...

1. The [`/base:` base address](https://learn.microsoft.com/en-us/cpp/build/reference/base-base-address?view=msvc-170) `link.exe` option is fixed for OCaml executable to be `0x10000` at https://github.com/ocaml/flexdll/blob/80496b51b9be9e128c6144377d20ccf6eff10129/cmdline.ml#L47
and also the same in
https://github.com/ocaml/flexdll/blob/80496b51b9be9e128c6144377d20ccf6eff10129/create_dll.ml#L77
for OCaml plugins / stubs.
2. All of the relative relocations from this section of code are translated by:
https://github.com/ocaml/flexdll/blob/80496b51b9be9e128c6144377d20ccf6eff10129/reloc.ml#L470-L475
into absolute relocations by `Reloc.abs`: https://github.com/ocaml/flexdll/blob/80496b51b9be9e128c6144377d20ccf6eff10129/coff.ml#L437-L444

So my suspicion was that the `Reloc.abs` absolute relocations were being translated at CreateProcess time by `ntdll.LdrInitializeThunk` (or whatever is reading the PE `.reloc` section) to complete the `imageBase` adjustment.

@dra27, any chance you can explain why `base_addr` (or `image_base` for DLLs) is not used in the calculations?

(These calculations are undocumented in flexdll and fairly complicated)

Reply:
It's far too long since I stared at all this, but are we comparing like with like here? Isn't that code in LLVM the part of a _linker_ which is computing actual relocations to be carried out by the _loader_? In flexdll's, case, I don't think we have to worry about the image base - the symbol table as a data structure has already been had all the relocations resolved when the executable was loaded (by the Windows loader) - so `s` in flexdll is an absolute address for the symbol, where in lld it's still base-relative?

Reply:
Yes, your loader comment is what I had meant with "translated at CreateProcess time by `ntdll.LdrInitializeThunk`".  But I am still puzzled why that only applies to IMAGE_REL_AMD64_ADDR64.